### PR TITLE
Adds support for subjective questions

### PR DESCRIPTION
**Pull request by dalmia**
## Summary
- adds an InputText component which supports a title, validation for the input in the text and a slot
- adds a Textarea component which uses the InputText component above and adds a `<textarea />` in its slot
- allows subjective as another question type

## Test Plan

- [x] Test Responsiveness
  - [ ] Laptop (1200px)
  - [ ] Tablet (760px)
  - [ ] Phone (320px)
- [x] Cross-Browser Testing
  - [ ] Chrome
  - [ ] Firefox
- [ ] ~Local Language Support~
- [ ] Hygiene and Housekeeping
  - [x] Self-review
  - [x] Comments have been added appropriately
  - [ ] ~Check for bundle size [here](https://bundlephobia.com/) if adding a package~
  - [x] Added relevant details like Labels/Projects/Milestones etc.
  - [ ] ~If adding or removing any environment variable:~
    - [ ] update `docs/ENV.md`
    - [ ] update Github Workflow files
    - [ ] update the secrets for staging and production
- [ ] Testing
  - [x] Wrote tests
  - [x] Tested locally
  - [x] Tested on staging
  - [ ] Tested on an actual physical phone
  - [ ] Tested on production
- [x] Lighthouse Checks
  - [ ] ~Images have `alt` attributes~
  - [ ] ~Any `` tags have `width` and `height` specified~
  - [ ] ~Any `target="_blank"` links have `rel="noopener"`~
  - [x] Only SVGs are used as images. If PNGs are used, their size has been optimised.
  - [ ] ~Any SVGs without text have their `aria-label` attributes set~

### CODE REVIEW

File: src/components/Questions/Body.vue [hunk 1]
```diff
@@ -70,25 +70,25 @@
...
+        <!-- input area for the answer -->
+        <Textarea
           v-model:value="subjectiveAnswer"
           class="px-2 w-full"
-          :boxStyling="subjectiveAnswerBoxStyling"
-          :placeholder="subjectiveAnswerInputPlaceholder"
-          :isDisabled="isAnswerSubmitted || previewMode"
-          :maxHeightLimit="subjectiveBoxHeightLimit"
+          boxStyling="bp-420:h-20 sm:h-28 md:h-36 px-4 placeholder-gray-400 focus:border-gray-200 focus:ring-primary"
+          placeholder="Enter your answer here"
```
Comment (pritamps) [inline src/components/Questions/Body.vue:84]: Why hardcoded strings? This won't mess with localization?

File: src/components/Questions/Body.vue [hunk 2]
```diff
@@ -70,25 +70,25 @@
...
+      >
+        <!-- input area for the answer -->
+        <Textarea
           v-model:value="subjectiveAnswer"
           class="px-2 w-full"
-          :boxStyling="subjectiveAnswerBoxStyling"
-          :placeholder="subjectiveAnswerInputPlaceholder"
-          :isDisabled="isAnswerSubmitted || previewMode"
-          :maxHeightLimit="subjectiveBoxHeightLimit"
+          boxStyling="bp-420:h-20 sm:h-28 md:h-36 px-4 placeholder-gray-400 focus:border-gray-200 focus:ring-primary"
```
Comment (pritamps) [inline src/components/Questions/Body.vue:83]: Why this change from variable to hardcoded string here?

File: src/components/Questions/Body.vue [hunk 3]
```diff
@@ -70,25 +70,25 @@
...
+          placeholder="Enter your answer here"
+          :isDisabled="isAnswerSubmitted"
+          :maxHeightLimit="250"
```
Comment (pritamps) [inline src/components/Questions/Body.vue:86]: Same questions as above.

File: src/components/Questions/Body.vue [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <template>
   <div class="overflow-y-auto flex flex-col">
     <!-- question text -->
-    <div class="px-4 md:px-6">
+    <div class="mx-6 md:mx-10">
```
Comment (pritamps) [inline src/components/Questions/Body.vue:4]: Why this change?

File: src/components/Questions/Body.vue [hunk 5]
```diff
@@ -212,7 +189,13 @@ export default defineComponent({
 
     function optionBackgroundClass(optionIndex: Number) {
       // returns the background class for the option
```
Comment (pritamps) [inline src/components/Questions/Body.vue:191]: This comment is useless. It's just reading out the function name lol. The function looks like it's doing something more with the `if`. Can you expand the comment a bit?

File: src/components/Questions/Body.vue [hunk 6]
```diff
@@ -230,6 +213,7 @@ export default defineComponent({
     function isOptionMarked(optionIndex: Number) {
       return (
         props.draftAnswer != null &&
+        typeof props.draftAnswer != "string" &&
```
Comment (pritamps) [inline src/components/Questions/Body.vue:216]: Still getting the hang of TS maybe, but why the condition of `!=string`?

File: src/components/Questions/Body.vue [hunk 7]
```diff
@@ -246,6 +230,12 @@ export default defineComponent({
       state.isImageLoading = true;
     }
 
+    function checkCharLimit(event: KeyboardEvent) {
+      // checks if character limit is reached in case it is set
+      if (!hasCharLimit.value) return;
+      if (!charactersLeft.value) event.preventDefault();
```
Comment (pritamps) [inline src/components/Questions/Body.vue:236]: What does `preventDefault` do?

File: src/components/Questions/Body.vue [hunk 8]
```diff
@@ -305,6 +295,40 @@ export default defineComponent({
...
+      if (state.subjectiveAnswer == null) return 0;
+      return state.subjectiveAnswer.length;
+    });
+    const defaultAnswer = computed(() => {
+      // the default answer to be shown for the subjective question
+      if (
+        props.submittedAnswer != null &&
+        typeof props.submittedAnswer == "string"
+      ) {
+        return props.submittedAnswer;
```
Comment (pritamps) [inline src/components/Questions/Body.vue:323]: Why are you returning `submittedAnswer` when the variable name is `defaultAnswer?

File: src/components/Questions/Body.vue [hunk 9]
```diff
@@ -305,6 +295,40 @@ export default defineComponent({
...
+        return "text-yellow-500";
+      } else return "text-red-400";
+    });
+    const charactersLeft = computed(() => {
+      // number of characters left for the subjective answer if a limit is given
+      return props.maxCharLimit - currentAnswerLength.value;
+    });
+    const currentAnswerLength = computed(() => {
+      // length of the current answer (for subjective question)
+      if (state.subjectiveAnswer == null) return 0;
```
Comment (pritamps) [inline src/components/Questions/Body.vue:314]: For non subjective answers, this variable *should* be `false` no, not `null`?

File: src/components/Questions/Body.vue [hunk 10]
```diff
@@ -246,6 +230,12 @@ export default defineComponent({
       state.isImageLoading = true;
     }
 
+    function checkCharLimit(event: KeyboardEvent) {
+      // checks if character limit is reached in case it is set
+      if (!hasCharLimit.value) return;
```
Comment (pritamps) [inline src/components/Questions/Body.vue:235]: Return for a function called `checkCharLimit` should be `true` or `false` and the caller should decide what to do with this info. Here, this function is deciding what to do, which may not be good in a case where there are multiple use cases for a character limit

File: src/components/Questions/QuestionModal.vue [hunk 11]
```diff
@@ -148,6 +151,11 @@ export default defineComponent({
       state.localCurrentQuestionIndex -= 1;
     }
 
+    /** update the attempt to the current question - valid for subjective questions */
```
Comment (pritamps) [inline src/components/Questions/QuestionModal.vue:154]: What is this doing? Allowing editing of subjective answers? Why is updating not valid for non-subjective questions?

File: src/components/UI/Icons/Correct.vue [hunk 12]
```diff
@@ -0,0 +1,21 @@
+<template>
```
Comment (pritamps) [inline src/components/UI/Icons/Correct.vue:1]: Does this belong in this PR? Adding correct/wrong applies to all types of questions, no?

File: src/components/UI/Icons/Wrong.vue [hunk 13]
```diff
@@ -0,0 +1,23 @@
+<template>
```
Comment (pritamps) [inline src/components/UI/Icons/Wrong.vue:1]: Does this belong in this PR? Adding correct/wrong applies to all types of questions, no?

File: src/views/Player.vue [hunk 14]
```diff
@@ -52,6 +52,15 @@ export default defineComponent({
           survey: false,
           max_char_limit: null,
         },
+        {
+          type: "subjective",
+          text: "yolo",
```
Comment (pritamps) [inline src/views/Player.vue:57]: :D

**Review by pritamps — CHANGES_REQUESTED** [review]
Mostly gaps in my understanding, I think. Maybe some more comments are needed along with a conversation about the marked questions

File: src/components/Questions/Body.vue (hunk 1 above)
Comment (dalmia) [reply src/components/Questions/Body.vue:84]: when we add support for internationalization, the value can be added directly here.

**Review by pritamps — APPROVED** [review]
Looks good to me!
